### PR TITLE
fulfill

### LINKED ISSUE #225: Admin - Fulfill Pickup Event Items

### PR DESCRIPTION
**Pull request by SheepTester**
# Info

Closes #225 

fulfills request for fulfill button

## Changes

- fulfill button for each order
- checkboxes for items that can be deselected to mark as not fulfilled
- got rid of fulfill view

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) 

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] on the live deployment preview on Desktop.
- [ ] on the live deployment preview on Mobile.
- [ ] I have added new Cypress tests that are passing.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented any new functions in `/src/lib/*` and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

image

image

### CODE REVIEW

File: src/components/admin/store/PickupOrder/index.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+              try {
+                const items = order.items.filter(item => !unselected.has(item.uuid));
+                const newOrder = await StoreAPI.fulfillOrderPickup(token, order.uuid, items);
+                const itemUuids = items.map(item => item.uuid);
+                onOrderUpdate({
+                  ...newOrder,
+                  items: order.items.map(item =>
+                    itemUuids.includes(item.uuid) ? { ...item, fulfilled: true } : item
+                  ),
+                });
```
Comment (alexzhang1618) [inline src/components/admin/store/PickupOrder/index.tsx]: seeking to learn: does the API response not just return the whole new order updated correctly? do we need to manually set fulfilled for the items we fulfilled

**Review by alexzhang1618 — COMMENTED** [review]
Code looks good! Good call on removing the order fulfill view i was just parroting off of v1 even though we don't really need it

Some usability comments:

can we put some gap between the status and fulfill button and maybe make the button the same size as the fulfill tag
Screenshot 2024-05-10 at 3 37 34 PM

also i think it would be nice to sort by status (i.e. put orders that still need action at the top)

I think we discussed this but I think ux wise it makes sense for all checkboxes to start blank and the merch person will tick each box off as they get the item

Another thing to consider: is it better to put the fulfill button under the merch items for each user? I think it's nicer so your mouse only has to stay on one side but idk

File: src/components/admin/store/PickupOrder/index.tsx (hunk 1 above)
Comment (SheepTester) [reply src/components/admin/store/PickupOrder/index.tsx]: the API response doesn't return the `items` array so it won't say which items were fulfilled

**Comment by SheepTester** [comment]
I changed the orders view to use badges instead. not sure if it's better

image


Marcelo cancelled this rescheduled partially fulfilled order, which is why it has one fulfilled item:

image

File: src/components/admin/store/PickupOrder/index.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        <ul className={styles.itemList}>
+          {itemQuantities.map(item => {
+            let badge = null;
+            if (
+              order.status === OrderStatus.FULFILLED ||
+              order.status === OrderStatus.PARTIALLY_FULFILLED
+            ) {
+              if (!item.fulfilled) {
+                badge = <span className={styles.notFulfilled}>Not fulfilled</span>;
+              }
```
Comment (alexzhang1618) [inline src/components/admin/store/PickupOrder/index.tsx]:
```suggestion
            if (
              !item.fulfilled && (
                order.status === OrderStatus.FULFILLED || 
                order.status === OrderStatus.PARTIALLY_FULFILLED
              )
            ) {
                badge = <span className={styles.notFulfilled}>Not fulfilled</span>;
```

File: src/components/store/OrderSummary/index.tsx [hunk 3]
```diff
@@ -57,6 +49,12 @@ const OrderItemPreview = ({ item, orderStatus }: OrderItemPreviewProps) => {
           alt="Store item picture"
           fill
         />
+        {showFulfilled && item.fulfilled ? (
+          <div className={styles.badge}>{showNotFulfilled ? 'Picked up' : 'Already picked up'}</div>
+        ) : null}
+        {showNotFulfilled && !item.fulfilled ? (
+          <div className={`${styles.badge} ${styles.notAvailable}`}>Not available</div>
```
Comment (alexzhang1618) [inline src/components/store/OrderSummary/index.tsx]: Not available makes me think that it's not in stock, can we do something like not picked up maybe? Could also ask others on what wording is best

File: src/components/store/OrderSummary/index.tsx [hunk 4]
```diff
@@ -57,6 +49,12 @@ const OrderItemPreview = ({ item, orderStatus }: OrderItemPreviewProps) => {
           alt="Store item picture"
           fill
         />
+        {showFulfilled && item.fulfilled ? (
+          <div className={styles.badge}>{showNotFulfilled ? 'Picked up' : 'Already picked up'}</div>
```
Comment (alexzhang1618) [inline src/components/store/OrderSummary/index.tsx]: Thoughts on like a green for this instead of grey similar to how notAvailable is red?

**Review by alexzhang1618 — COMMENTED** [review]
I like the tags and the extra text, it looks nice!
Some UI comments:

I think I prefer ✅ and ❌ emojis to the text since it's more concise and readable

I'm also not sure how much I like having all these check boxes to the right of the item. In our store, we don't really have a lot of items where people are ordering a big quantity of them, but maybe we should make this just one check to fulfill all or maybe a number to fulfill a specific number of them? Up to you and whatever looks best
Screenshot 2024-05-18 at 12 26 44 PM

File: src/components/store/OrderSummary/index.tsx (hunk 3 above)
Comment (SheepTester) [reply src/components/store/OrderSummary/index.tsx]: yeah it should only show for partially fulfilled events, so it wasn't in stock

image

**Comment by SheepTester** [comment]
- I switched the text labels in the admin orders page back to emoji
- I merged all the checkboxes into one checkbox per variant. This means that if someone orders multiple ACM AI stickers, but there's somehow only one sticker left, you can only either mark them all as fulfilled or not fulfilled. Maybe that's fine for now.
- I changed the "picked up" / "already picked up" (for rescheduled orders) badge to green:
  
  image

File: src/components/admin/store/PickupOrder/index.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                onOrderUpdate({
+                  ...newOrder,
+                  items: order.items.map(item =>
+                    itemUuids.includes(item.uuid) ? { ...item, fulfilled: true } : item
+                  ),
+                });
+              } catch (error: unknown) {
+                reportError('Failed to fulfill order', error);
+              }
+            }}
```
Comment (alexzhang1618) [inline src/components/admin/store/PickupOrder/index.tsx]: tiny nit but if there's a good amt of code here i'd rather it be a helper function `handleFulfillOrder` so we can keep the jsx nice and tidy

**Review by alexzhang1618 — APPROVED** [review]
LGTM, some nits

Maybe a word break and a little margin between the checkbox n the button?
Screenshot 2024-08-19 at 11 01 14 PM

**Comment by SheepTester** [comment]
image

**Pull request merged**